docs/source/notes/large_scale_deployments.rst

```diff
@@ -0,0 +1,136 @@
+context (e.g. `name`). If enabled, ``RecordFunction::inputs()`` contains arguments
+of the function represented as ``torch::IValue`` variant type. Note, that inputs
+logging is relatively expensive and thus has to be enabled explicitly.
+
+Invoking callbacks adds some overhead, so usually it's useful to just randomly
+sample operator invocations. This can be enabled on per-callback basis with a
+global sampling rate specified by
+`torch::autograd::profiler::setSamplingProbability`.
+
+Note, that ``pushCallback`` and ``setSamplingProbability`` are not thread-safe
+and can be called only when no PyTorch operator is running. Usually, it's a good
+idea to call them once during initialization.
+
+Here's an example:
+
+.. code-block:: cpp
+
+    // Called somewhere in the program beginning
+    void init() {
+        // Sample one in a hundred operator runs randomly
+        torch::autograd::setSamplingProbability(0.01);
+        pushCallback(
+            &onFunctionEnter,
+            &onFunctionExit,
+            /* needs_inputs */ true,
+            /* sampled */ true
+        );
+    }
+
+    void onFunctionEnter(const RecordFunction& fn) {
+        std::cerr << "Before function " << fn.name() 
+                  << " with " << fn.inputs().size() << " inputs" << std::endl;
+    }
+
+    void onFunctionExit(const RecordFunction& fn) {
+        std::cerr << "After function " << fn.name();
+    }
+
+API usage logging
+^^^^^^^^^^^^^^^^^
+
+When running in a broader ecosystem, for example in managed job scheduler, it's
+often useful to track which binaries invoke particular PyTorch APIs. There
+exists simple instrumentation injected at several important API points that
+triggers a given callback. Because usually PyTorch is invoked in one-off python
+scripts, the callback fires only once for a given process for each of the APIs.
+
+``c10::SetAPIUsageHandler`` can be used to register API usage instrumentation
+handler. Passed argument is going to be an "api key" identifying used point, for
+example ``python.import`` for PyTorch extension import or
+``torch.script.compile`` if TorchScript compilation was triggered.
+
+.. code-block:: cpp
+
+    SetAPIUsageLogger([](const std::string& event_name) {
+        std::cerr << "API was used: " << event_name << std::endl;
+    });
+
+Note for developers: new API trigger points can be added in code with
+``C10_LOG_API_USAGE_ONCE("my_api")`` in C++ or
+``torch._C._log_api_usage_once("my.api")`` in Python.
+
+Attaching metadata to saved TorchScript models
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+TorchScript modules can be saved as an archive file that bundles serialized
+parameters and module code as TorchScript (see :meth:`torch.jit.save`). It's
+often convenient to bundle additional information together with the model, for
+example, description of model producer or auxiliary artifacts.
+
+It can be achieved by passing the ``_extra_files`` argument to
```

Comment:
Well we should never document arguments which have an underscore prefix... It's there to signify it's private. This has been put as a temporary measure after a discussion I had with @zdevito, but we never intended this functionality to be publicized and to have a lot of people depend on it...

Reply:
Sorry, I missed this comment. I think passing metadata through PT files might be useful for higher level workflows built on top of PT. How do you feel about dropping underscore all together and making it official functionality? I don't see too many drawbacks as the interface name->blob is pretty generic and our container is a zip file already.

Of course, one can argue that additional files can be appended with just `ZipFile` python API. While it's ok for reading, there might be unintended effects for writing as ZipFile doesn't guarantee data block alignment (something our writer does). For reading, suggesting to use ZipFile should be sufficient.